pkg/providers/vsphere/network/reconcile.go

```diff
@@ -0,0 +1,151 @@
+// © Broadcom. All Rights Reserved.
+// The term “Broadcom” refers to Broadcom Inc. and/or its subsidiaries.
+// SPDX-License-Identifier: Apache-2.0
+
+package network
+
+import (
+	"context"
+	"strings"
+
+	"github.com/vmware/govmomi/object"
+	vimtypes "github.com/vmware/govmomi/vim25/types"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	netopv1alpha1 "github.com/vmware-tanzu/net-operator-api/api/v1alpha1"
+	vpcv1alpha1 "github.com/vmware-tanzu/nsx-operator/pkg/apis/vpc/v1alpha1"
+	ncpv1alpha1 "github.com/vmware-tanzu/vm-operator/external/ncp/api/v1alpha1"
+
+	"github.com/vmware-tanzu/vm-operator/pkg/util/resize"
+)
+
+func ReconcileNetworkInterfaces(
+	ctx context.Context,
+	results *NetworkInterfaceResults,
+	currentEthCards object.VirtualDeviceList,
+) ([]vimtypes.BaseVirtualDeviceConfigSpec, error) {
+
+	var deviceChanges []vimtypes.BaseVirtualDeviceConfigSpec
+
+	for idx, r := range results.Results {
+		matchingIdx := FindMatchingEthCard(currentEthCards, r.Device.(vimtypes.BaseVirtualEthernetCard))
+		if matchingIdx >= 0 {
+			// Exact match. Claim it by removing the device from the current ethernet cards.
+			matchDev := currentEthCards[matchingIdx].(vimtypes.BaseVirtualEthernetCard).GetVirtualEthernetCard()
+			results.Results[idx].DeviceKey = matchDev.Key
+			results.Results[idx].MacAddress = matchDev.MacAddress
+			currentEthCards = append(currentEthCards[:matchingIdx], currentEthCards[matchingIdx+1:]...)
+		} else {
+			existingIdx := findExistingEthCardForOrphanedCR(ctx, r.Name, results.OrphanedNetworkInterfaces, currentEthCards)
+			if existingIdx >= 0 {
+				editDev := currentEthCards[existingIdx]
+
+				ethDev := editDev.(vimtypes.BaseVirtualEthernetCard).GetVirtualEthernetCard()
+				ethDev.Backing = r.Device.GetVirtualDevice().Backing
+				ethDev.AddressType = r.Device.(vimtypes.BaseVirtualEthernetCard).GetVirtualEthernetCard().AddressType
+				ethDev.MacAddress = r.Device.(vimtypes.BaseVirtualEthernetCard).GetVirtualEthernetCard().MacAddress
+				ethDev.ExternalId = r.Device.(vimtypes.BaseVirtualEthernetCard).GetVirtualEthernetCard().ExternalId
+
+				deviceChanges = append(deviceChanges, &vimtypes.VirtualDeviceConfigSpec{
+					Device:    editDev,
+					Operation: vimtypes.VirtualDeviceConfigSpecOperationEdit,
+				})
+
+				currentEthCards = append(currentEthCards[:existingIdx], currentEthCards[existingIdx+1:]...)
+			} else {
+				deviceChanges = append(deviceChanges, &vimtypes.VirtualDeviceConfigSpec{
+					Device:    r.Device,
+					Operation: vimtypes.VirtualDeviceConfigSpecOperationAdd,
+				})
+			}
+
+			results.UpdatedEthCards = true
+		}
+	}
+
+	// Remove any unmatched existing interfaces.
+	removeDeviceChanges := make([]vimtypes.BaseVirtualDeviceConfigSpec, 0, len(currentEthCards))
+	for _, dev := range currentEthCards {
+		removeDeviceChanges = append(removeDeviceChanges, &vimtypes.VirtualDeviceConfigSpec{
```

Comment:
Used `make` w/ cap, but appending here. Using `removeDeviceChanges[i] = ...` instead?

Reply:
This ends up tripping another lint warning wrt the append in the return statement so I think I'll just leave it as-is